keps/sig-node/3542-cri-image-pulling-with-progress-notification/README.md

```diff
@@ -0,0 +1,876 @@
+
+### Goals
+
+- Extend CRI to provide image pulling progress API that can be utilized by client
+tools (crictl pull), or machinery (kubelet pulling the image)
+- Extend existing PullImage api to support a duration timeout with no progress feature
+- Implement PoC / draft for CRI PullImageWithProgress call to runtime in Kubelet, hidden behind
+FeatureGate, disable by default.
+
+### Non-Goals
+
+- Complete implementation of utilization of proposed interface additions in the client-tools and / or runtimes
+
+## Proposal
+
+We propose:
+
+- extending existing ImagePull call with optional parameter:
+  - no-progress timeout: number of seconds during which if completely no data transfer was ongoing, failure should be reported
+- introducing new (additional, not replacing old one) API for requesting image pull,
+that will return stream with periodic updates sent through it to the client until completion or
+a timeout without download progress is reached. The image pull with progress request parameters will contain:
+  - image pull request data structure, including new optional no-progress timeout field
+  - the type of granularty based on which the client wants to receive updates about the progress:
+    - time-based
+    - size-based
+    - none
+  - interval of the updates (if any) respectively to the granularity type:
+    - N seconds
+    - N amount of data downloaded of the total image size, e.g. 1Gi
+  - verbosity
+    - summarized - single value, total current download progress against total estimated download size
+    - verbose - summarized plus additionally per layer information of download progress against layer size
+
+If / when it is possible to reliably determine percentage of the progress in the runtime,
+percentage-based granularity type can be introduced then.
+
+If the client did not specify the preferred notification granularity, default values should be used,
+for instance every Gibibyte downloaded or every 60 seconds of time spent downloading an image.
+
+
+### Kubelet defaults
+
+For Kubelet implementation we propose:
+- for alpha: disabling image pulling progress messages in request to Runtime
+- default no-progress timeout to be 10 seconds
+
+Example Kubelet image pull with progress request:
+```
+{
+    request: {...},
+    granularity_type: "none",
+    interval: 0,
+    no_progress_timeout: 10,
+    verbosity: false,
+}
+```
+
+### Kubelet config
+
+The granularity type, interval, and no-progress timeout should be configurable through kubelet-config.
+
+Suggested new Kubelet config fields are these:
+
+       // NoProgressTimeout is a number of seconds after which stalled image download should be reported
+       // as an error.
+       // Supported values are:
+       // - 0 for infinity, effectively no timeout
+       // - positive number up to 4294967295 (uint32 max number, approx. 136 years)
+       // Default: 10
+       // +optional
+       NoProgressTimeout int32 `json:"NoProgressTimeout,omitempty"`
```

Comment:
How will this interact with the current CRI timeout?

Reply:
I suppose this was meant to be much shorter than CRI timeout and tell the runtime how soon to report failure if there was no progress in image downloading. Without it the CRI timeout would cancel the call much later.